core/src/main/java/kafka/server/NetworkPartitionMetadataClient.java

```diff
@@ -0,0 +1,305 @@
+                tp.topic(),
+                tp.partition(),
+                listenerName
+            );
+
+            if (leaderNodeOpt.isEmpty() || leaderNodeOpt.get().isEmpty()) {
+                // No leader available - complete with error
+                futures.put(tp, CompletableFuture.completedFuture(new OffsetResponse(-1, Errors.LEADER_NOT_AVAILABLE)));
+                continue;
+            }
+
+            partitionsByNode.computeIfAbsent(leaderNodeOpt.get(), k -> new ArrayList<>()).add(tp);
+        }
+
+        // Create and enqueue requests for each node
+        partitionsByNode.forEach((node, partitionsByLeader) -> {
+            // All partitions with the same leader node will be included in the same ListOffsetsRequest.
+            Map<TopicPartition, CompletableFuture<OffsetResponse>> partitionFuturesByLeader = new HashMap<>();
+            for (TopicPartition tp : partitionsByLeader) {
+                CompletableFuture<OffsetResponse> future = new CompletableFuture<>();
+                futures.put(tp, future);
+                partitionFuturesByLeader.put(tp, future);
+            }
+
+            // Create ListOffsetsRequest for this node
+            ListOffsetsRequest.Builder requestBuilder = createListOffsetsRequest(partitionsByLeader);
+            // Create pending request to track this request
+            PendingRequest pendingRequest = new PendingRequest(node, partitionFuturesByLeader, requestBuilder);
+            // Enqueue to send thread
+            sendThread.enqueue(pendingRequest);
+        });
+
+        return futures;
+    }
+
+    @Override
+    public void close() {
+        // Only close sendThread if it was initialized. Note, close is called only during broker shutdown, so need
+        // for further synchronization here.
+        if (!initialized.get()) {
+            return;
+        }
+        if (sendThread != null) {
+            try {
+                sendThread.shutdown();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                log.error("Interrupted while shutting down NetworkPartitionMetadataClient", e);
+            }
+        }
+    }
+
+    /**
+     * Ensures that the sendThread is initialized. This method is thread-safe and will only
+     * initialize the sendThread once, even if called concurrently.
+     */
+    // Visible for testing.
+    void ensureSendThreadInitialized() {
+        if (initialized.compareAndSet(false, true)) {
+            KafkaClient networkClient = networkClientSupplier.get();
+            sendThread = new SendThread(
+                "NetworkPartitionMetadataClientSendThread",
+                networkClient,
+                Math.toIntExact(CommonClientConfigs.DEFAULT_SOCKET_CONNECTION_SETUP_TIMEOUT_MAX_MS),  //30 seconds
+                this.time
+            );
+            sendThread.start();
+            log.info("NetworkPartitionMetadataClient sendThread initialized and started");
```

Comment:
I don't think we want an INFO log for this all the time. I see how it's useful while the code is being developed.